docs_nnx/key_concepts.md

```diff
@@ -0,0 +1,320 @@
+
+
+```python
+def jax_linear(x, kernel, bias):
+  return jnp.dot(x, kernel) + bias
+
+params = {'kernel': jax.random.normal(jax.random.key(42), (4, 2)), 
+          'bias': jnp.zeros((2,))}
+x = jax.random.normal(jax.random.key(0), (2, 4))
+y = jax_linear(x, params['kernel'], params['bias'])
+```
+
+## What is Flax?
+
+Flax is a **neural network toolkit**, offering higher level abstractions that are handy for model developers. Such as:
+
+* **Object-oriented `Module` class** to represent layers/models and bookkeep parameters.
+
+* **Modeling utilities** like random number handling, model traversal and surgery, optimizers, advanced parameter bookkeeping, sharding annotations, and more.
+
+* **Some built-in commonly-used** layers, initializers, and model examples.
+
+Take the example below: A Flax layer `Linear`, during initialization, takes one RNG key and automatically initialize all internal parameters as `jax.Array`s. In forward pass, it carries out the exact same computation via JAX APIs.
+
+
+```python
+# Eligible parameters were created inside `linear`, using one RNG key 42
+linear = nnx.Linear(in_features=4, out_features=2, rngs=nnx.Rngs(42))
+
+# Flax created a `Param` wrapper over the actual `jax.Array` parameter to track metadata
+print(type(linear.kernel))        # flax.nnx.Param
+print(type(linear.kernel.value))  # jax.Array
+
+# The computation of the two are the same
+x = jax.random.normal(jax.random.key(0), (2, 4))
+flax_y = linear(x)
+jax_y = jax_linear(x, linear.kernel.value, linear.bias.value)
+assert jnp.array_equal(flax_y, jax_y)
+```
+
+    <class 'flax.nnx.variablelib.Param'>
+    <class 'jaxlib._jax.ArrayImpl'>
+
+
+## Pytrees
+
+Your code likely needs more than one `jax.Array`. A **pytree** is a container structure of multiple pytrees, possibly nested. It is a key and handly concept in the JAX world.
+
+Many things are pytrees: Python dicts, lists, tuples, dataclasses, and more. The key is that a pytree can be "flattened" into multiple children, which are either pytrees or individual `jax.Array`s. JAX allows you to  register your own classes as pytree nodes.
+
+Pytree is the primary data holder in JAX. When JAX transforms see a pytree argument, they automatically trace its internal `jax.Array`s when compiling. Therefore, it's crucial to organize your data as pytrees. [JAX pytree documentation](https://docs.jax.dev/en/latest/working-with-pytrees.html) has a thorough overview on pytrees and JAX APIs to manipulate them.
+
+In Flax, a `Module` is not a pytree, but its state is a pytree. This how we implement Flax transforms (e.g. `flax.nnx.vmap` instead of `jax.vmap`) - the model is first converted to a JAX-compatible pytree dict, then the underlying JAX transforms are called.
+
+In the future, Flax Modules will also become pytrees, allowing us to directly using JAX transforms.
```

Comment:
The future is now! 😅 (#4811) Should we update this section?
Maybe we can merge this an do an update.

Reply:
Aha! I quickly updated it to briefly show the latest fact. We can have more elaborate and examples later.